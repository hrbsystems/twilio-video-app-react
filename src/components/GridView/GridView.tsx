import React from 'react';
import { GRID_MODE_ASPECT_RATIO, GRID_MODE_MARGIN, GRID_MODE_MAX_PARTICIPANTS } from '../../constants';
import { makeStyles, Theme } from '@material-ui/core';
import Participant from '../Participant/Participant';
import useGridLayout from '../../hooks/useGridLayout/useGridLayout';
import useParticipants from '../../hooks/useParticipants/useParticipants';
import useVideoContext from '../../hooks/useVideoContext/useVideoContext';

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    display: 'flex',
    width: 'calc(100% - 200px)',
    margin: '0 auto',
    alignContent: 'center',
    flexWrap: 'wrap',
    justifyContent: 'center',
    gridArea: '1 / 1 / 2 / 3',
  },
  participant: {
    '&:nth-child(n + 26)': {
      display: 'none',
    },
  },
}));

export function GridView() {
  const classes = useStyles();
  const { room } = useVideoContext();
  const participants = useParticipants();

  const { participantVideoWidth, containerRef } = useGridLayout(
    Math.min(participants.length + 1, GRID_MODE_MAX_PARTICIPANTS)
  );

  const participantWidth = `${participantVideoWidth}px`;
  const participantHeight = `${Math.floor(participantVideoWidth * GRID_MODE_ASPECT_RATIO)}px`;

  return (
    <div className={classes.container} ref={containerRef}>
      <div
        className={classes.participant}
        style={{ width: participantWidth, height: participantHeight, margin: GRID_MODE_MARGIN }}
      >
        <Participant participant={room!.localParticipant} isLocalParticipant={true} />
      </div>
      {participants.map(participant => (
        <div
          key={participant.sid}
          className={classes.participant}
          style={{ width: participantWidth, height: participantHeight, margin: GRID_MODE_MARGIN }}
        >
          <Participant participant={participant} />
        </div>
      ))}
    </div>
  );
}
