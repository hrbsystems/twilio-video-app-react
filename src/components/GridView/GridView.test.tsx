import React from 'react';
import { GridView } from './GridView';
import { shallow } from 'enzyme';
import useGridLayout from '../../hooks/useGridLayout/useGridLayout';

const mockParticipants = [
  { identity: 'test-participant-1', sid: 1 },
  { identity: 'test-participant-2', sid: 2 },
  { identity: 'test-participant-3', sid: 3 },
  { identity: 'test-participant-4', sid: 4 },
];

jest.mock('../../constants', () => ({
  GRID_MODE_ASPECT_RATIO: 9 / 16,
  GRID_MODE_MAX_PARTICIPANTS: 2,
  GRID_MODE_MARGIN: 3,
}));
jest.mock('../../hooks/useParticipants/useParticipants', () => () => mockParticipants);
jest.mock('../../hooks/useVideoContext/useVideoContext', () => () => ({
  room: {
    localParticipant: { identity: 'test-local-participant' },
  },
}));
jest.mock('../../hooks/useGridLayout/useGridLayout', () =>
  jest.fn(() => ({
    participantVideoWidth: 720,
    containerRef: { current: null },
  }))
);

describe('the GridView component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<GridView />);
    expect(wrapper).toMatchSnapshot();
    expect(useGridLayout).toHaveBeenCalledWith(2);
  });
});
