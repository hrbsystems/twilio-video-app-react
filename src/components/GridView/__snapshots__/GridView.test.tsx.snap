// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the GridView component should render correctly 1`] = `
<div
  className="makeStyles-container-1"
>
  <div
    className="makeStyles-participant-2"
    style={
      Object {
        "height": "405px",
        "margin": 3,
        "width": "720px",
      }
    }
  >
    <Participant
      isLocalParticipant={true}
      participant={
        Object {
          "identity": "test-local-participant",
        }
      }
    />
  </div>
  <div
    className="makeStyles-participant-2"
    key="1"
    style={
      Object {
        "height": "405px",
        "margin": 3,
        "width": "720px",
      }
    }
  >
    <Participant
      participant={
        Object {
          "identity": "test-participant-1",
          "sid": 1,
        }
      }
    />
  </div>
  <div
    className="makeStyles-participant-2"
    key="2"
    style={
      Object {
        "height": "405px",
        "margin": 3,
        "width": "720px",
      }
    }
  >
    <Participant
      participant={
        Object {
          "identity": "test-participant-2",
          "sid": 2,
        }
      }
    />
  </div>
  <div
    className="makeStyles-participant-2"
    key="3"
    style={
      Object {
        "height": "405px",
        "margin": 3,
        "width": "720px",
      }
    }
  >
    <Participant
      participant={
        Object {
          "identity": "test-participant-3",
          "sid": 3,
        }
      }
    />
  </div>
  <div
    className="makeStyles-participant-2"
    key="4"
    style={
      Object {
        "height": "405px",
        "margin": 3,
        "width": "720px",
      }
    }
  >
    <Participant
      participant={
        Object {
          "identity": "test-participant-4",
          "sid": 4,
        }
      }
    />
  </div>
</div>
`;
